ops/pebble.py

```diff
@@ -1054,6 +1045,29 @@ def _make_auth_dict(permissions, user_id, user, group_id, group) -> typing.Dict:
             d['group'] = group
         return d
 
+    @staticmethod
+    def _encode_multipart(metadata, path, content):
+        # Python's stdlib mime/multipart handling is screwy and doesn't handle
+        # binary properly, so roll our own.
+        boundary = binascii.hexlify(os.urandom(16))
+        path_escaped = path.replace('"', '\\"').encode('utf-8')  # NOQA: test_quote_backslashes
+        parts = []
+        parts.extend([
+            b'--', boundary, b'\r\n',
+            b'Content-Type: application/json\r\n',
+            b'Content-Disposition: form-data; name="request"\r\n',
+            b'\r\n',
+            json.dumps(metadata).encode('utf-8'), b'\r\n',
+            b'--', boundary, b'\r\n',
+            b'Content-Type: application/octet-stream\r\n',
+            b'Content-Disposition: form-data; name="files"; filename="', path_escaped, b'"\r\n',
+            b'\r\n',
+            content, b'\r\n',
+            b'--', boundary, b'--\r\n',
+        ])
+        content_type = 'multipart/form-data; boundary="' + boundary.decode('utf-8') + '"'
+        return b''.join(parts), content_type
```

Comment:
So this fundamentally buffers everything, including the content in memory. It actually buffers everything 2x, because you have the 'content' string, and then the ''.join() string.
I thought that was the hope of multipart that we could move to some sort of streaming/chunked implementation.

https://github.com/python/cpython/blob/main/Lib/urllib/request.py#L1277

Says that Request.data can be a file-like object, rather than being the raw bytes.
And it further has a check that if there isn't a Content-Length that it knows about, it can explicitly turn the stream into Transfer-encoding: chunked.

That said, it is unclear to me how multipart and transfer-encoding: chunked play together. But I think if we just made parts have a '.read' method we could get away from buffering all of the bytes in memory.

Thoughts?

Reply:
Yeah, you're not wrong -- however, this is the same as [the previous code](https://github.com/canonical/operator/pull/574/files#diff-d74b54ce28b898e8febb0f33f24abc9bf91bb7793c51ae716bf9a9816d90104dL1032), which loaded the whole file into memory (x2 with the `as_bytes()` call) to keep the initial implementation simple. So this PR simply fixes the binary corruption bug; it doesn't change the memory characteristics.

My plan was to come back as needed and improve that (and the API was intentionally designed such that it supports a file-like object, so no change will be needed to the API). I've opened https://github.com/canonical/operator/issues/576 to track.

Regarding `Transfer-Encoding: chunked` -- yeah, that works fine with multipart. The transfer encoding is applied at a low level and just allows the content to be streamed in chunks, for example one multipart section flushed out at a time (or smaller chunks). Go handles this transparently; looks like Python might too given the right combination of inputs/headers.